packages/core/mesh/network-manager/src/transport/libdatachannel-transport.ts

```diff
@@ -0,0 +1,180 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Duplex } from 'stream';
+
+import { Event, Trigger } from '@dxos/async';
+import { ErrorStream } from '@dxos/debug';
+import { log } from '@dxos/log';
+import { Signal } from '@dxos/protocols/proto/dxos/mesh/swarm';
+
+import { type PeerConnection } from './datachannel/rtc-peer-connection';
+import { Transport } from './transport';
+
+export type LibDataChannelTransportParams = {
+  initiator: boolean;
+  stream: NodeJS.ReadWriteStream;
+  webrtcConfig?: RTCConfiguration;
+  sendSignal: (signal: Signal) => Promise<void>;
+};
+
+const MAX_BUFFERED_AMOUNT = 64 * 1024;
+
+/**
+ * Transport
+ */
+export class LibDataChannelTransport implements Transport {
+  readonly closed = new Event();
+  readonly connected = new Event();
+  readonly errors = new ErrorStream();
+  private readonly _peer: Promise<PeerConnection>;
+  private _channel!: RTCDataChannel;
+  private _stream!: Duplex;
+
+  private readonly _readyForCandidates = new Trigger();
+  private _writeCallback: (() => void) | null = null;
+
+  constructor(private readonly params: LibDataChannelTransportParams) {
+    this._peer = (async () => {
+      const { PeerConnection } = await import('./datachannel/rtc-peer-connection');
+      const peer = new PeerConnection(params.webrtcConfig);
```

Comment:
possible race condition - dont create a peer if the connection was closed by this point